crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -244,13 +236,132 @@ impl<'flox> Publish<'flox, NixAnalysis> {
     }
 }
 
+/// Representation of an exclusive clone of an upstream repo
+///
+/// [UpstreamRepo] and [UpstreamCatalog] ensure safe access to individual catalog branches.
+/// Every [UpstreamRepo] instance represents an exclusive clone
+/// and can only ever create a single [UpstreamCatalog] instance at a time.
+struct UpstreamRepo(Git);
+
+impl UpstreamRepo {
+    /// Clone the upstream repo
+    async fn clone_repo(
+        publish_ref: &PublishRef,
+        temp_dir: impl AsRef<Path>,
+    ) -> Result<Self, PublishError> {
+        let url = publish_ref.clone_url();
+        let repo_dir = tempfile::tempdir_in(temp_dir).unwrap().into_path(); // todo catch error
+        let repo = <Git as GitProvider>::clone(&url, &repo_dir, false).await?;
+
+        Ok(Self(repo))
+    }
+
+    fn catalog_branch_name(system: &System) -> String {
+        format!("catalog/{system}")
+    }
+
+    /// Create an [UpstreamCatalog] by checking out or creating a catalog branch.
+    ///
+    /// `Git` objects can switch branches at any time leaving the repo in an unknown state.
+    /// [get_catalog] ensures that only one [UpstreamCatalog] exists at a time by requiring a `&mut self`.
```

Comment:
Does that guarantee that a method in UpstreamCatalog doesn't call checkout?

Reply:
yes, `&mut` references are enforced to be unique by the compiler.

Because `UpstreamCatalog` holds a reference from a mutable reference on `UpstreamCatalog`, `UpstreamCatalog` can't be borrowed until the catalog with that reference is dropped.

The alternative is to use move semantics and move `UpstreamRepo` into the type, i.e. 

```diff 
- struct UpstreamCatalog<'a>(&'a Git);
+ struct UpstreamCatalog<'a>(&'a Git, UpstreamRepo);

- fn get_catalog(&mut self, system: &System) -> Result<UpstreamCatalog, PublishError>
+ fn get_catalog(self, system: &System) -> Result<UpstreamCatalog, PublishError>

```

Reply:
Sure if the implementation of `UpstreamCatalog` performs a checkout, that's nothing that is prevented by this.
The point of this guarantee is that you can't get two `UpstreamCatalog` instances of the same repo at the same time from an `UpstreamRepo`